### PR TITLE
pm: Declare pm state constraints for a device

### LINKED ISSUE #67481: PM policy: map device driver instances to power states disabling power to the devices
**The problem**

When a device driver activates a device, it should ensure that the power to the device won't be cut off. To do this, the device driver forbids the power management subsystem from entering the power states which would result in disabling power to the activated device. This can done with the following functions:

https://github.com/zephyrproject-rtos/zephyr/blob/7d7f7aaf03fd0728d298dce22c20cb3098c3d0aa/include/zephyr/pm/policy.h#L114
https://github.com/zephyrproject-rtos/zephyr/blob/7d7f7aaf03fd0728d298dce22c20cb3098c3d0aa/include/zephyr/pm/policy.h#L125

A device driver is expected to be a generic implementation for a given device. A single device may be integrated with different SoCs, by different vendors, or with different boards. Each of them can support other power states, and other power states may result in disabling power for this device. Moreover, in a single SoC, there can be multiple instances of a single device type (using the same driver) and each of these instances can be powered off in other power states.

The device driver implementation should be agnostic to the device integration environment, and because of this it shall not have hardcoded power state and sub-state values which are required in the `void pm_policy_state_lock_get(enum pm_state state, uint8_t substate_id);` function.

**The solution**

I would like to have a mapping between device instances and power states that disable power to a given device. I would like to express it in the device tree reusing already existing devices and power state nodes, like:

```
power-states {
                idle: system_on_idle {
                        compatible = "zephyr,power-state";
                        power-state-name = "suspend-to-idle";
                        min-residency-us = <100>;
                };

                suspend_to_ram: s2ram {
                        compatible = "zephyr,power-state";
                        power-state-name = "suspend-to-ram";
                        min-residency-us = <10000>;
                        exit-latency-us = <1000>;
                };
};

a_peripheral_bus {
        uart0: uart@8000 {
                compatible = "nordic,nrf-uarte";
                reg = <0x8000 0x1000>;
                interrupts = <8 NRF_DEFAULT_IRQ_PRIORITY>;
                power-states-disabling-power = <&suspend_to_ram>;
                status = "disabled";
        };

        uart1: uart@9000 {
                compatible = "nordic,nrf-uarte";
                reg = <0x9000 0x1000>;
                interrupts = <9 NRF_DEFAULT_IRQ_PRIORITY>;
                power-states-disabling-power = <>;  // This instance is powered on in all power states
                status = "disabled";
        };
};
```

With this mapping, the driver code could call a new power management function implemented as:

```
void pm_policy_device_power_lock_get(const struct *device)
{
    enum pm_state state;
    uint8_t substate_id;

    for (POWER_STATES_DISABLING_POWER_FOR_DEVICE(device, state, substate_id)) {
        pm_policy_state_lock_get(state, substate_id);
    }
}

void pm_policy_device_power_lock_put(const struct *device)
{
    enum pm_state state;
    uint8_t substate_id;

    for (POWER_STATES_DISABLING_POWER_FOR_DEVICE(device, state, substate_id)) {
        pm_policy_state_lock_put(state, substate_id);
    }
}
```

This way the driver gets decoupled from its device environment and the power tree in the product. The power tree is handled in the module dedicated to power management.

**The alternatives**
I considered using existing APIs, but it leads to hard to maintain macros and unnecessary coupling between the device drivers and the power tree in the products integrating the devices like:

```
#ifdef CONFIG_SOC_NRF...
    pm_policy_state_lock_get(1, 0);
    pm_policy_state_lock_get(1, 2);
    pm_policy_state_lock_get(2, 0);
#elif CONFIG_SOC_STM32...
    pm_policy_state_lock_get(0, 0);
#endif
```

I was also thinking about using "power domains" in the device tree, but I'm not familiar enough with this concept to include it in the proposal.

### PR DESCRIPTION
**Pull request by ceolin**
Declare in devicetree which power states would cause power loss in a device. This information can be used by devices
to set/release power state constraints to avoid being suspended when they are in the middle of an operation.

### CODE REVIEW

**Comment by ceolin** [comment]
@JordanYates @erwango @hubertmis please a take a look if you can. Any feedback is welcome, especially around naming convention.

**Comment by erwango** [comment]
On my side, I don't think it is a good approach to declare pm states in device nodes.
Interest of dts is when you need to be able to distinguish various behaviors of different instances in the same device driver (either due to devices specifics (IRQ number) or application specifics (speed, pins, ...). I don't think this could happen in current case as all instances are supposed to behave the same way and have the same constraints.
I some rare occasions (see https://github.com/zephyrproject-rtos/zephyr/pull/60369), this might be different, but it doesn't justify the need to have to populate supported PM states in each and every node implementing PM.

**Comment by ceolin** [comment]
> On my side, I don't think it is a good approach to declare pm states in device nodes. Interest of dts is when you need to be able to distinguish various behaviors of different instances in the same device driver (either due to devices specifics (IRQ number) or application specifics (speed, pins, ...). I don't think this could happen in current case as all instances are supposed to behave the same way (ie PM state support implemented or not) and have the same constraints. I some rare occasions (see #60369), this might be different, but it doesn't justify the need to have to populate supported PM states in each and every node implementing PM. Proposed approach is overkill to me.

It may be indeed. The problem is that is hard to make generic drivers without this information, but to be honest that is not being a source of complains. I was already thinking about making it protected under a build option to avoid waste memory.

But yeah, this increases the complexity and the number of customization for power management. The problem is that we have really different use cases to support. I don't see a way now to a driver, that can be used in different targets. to set constraints to avoid being power off without it. At the moment the driver needs to exactly know which power states the soc uses.

File: tests/subsys/pm/power_states_api/src/main.c
Comment (hubertmis) [inline tests/subsys/pm/power_states_api/src/main.c:95]: I think `zaasert(suspend_to_ram_count == 0);` should be here. Or am I missing something?

**Review by hubertmis — COMMENTED** [review]
Thank you, this looks like the missing feature I was looking for!

**Comment by hubertmis** [comment]
> Interest of dts is when you need to be able to distinguish various behaviors of different instances in the same device driver

That's exactly the reason to define power states in .dts. An example is nRF54H20 in which one instance of MVDMA gets power cut off in 2 sleep states while the remaining instances lose power in 3 sleep states. So the MVDMA driver needs to distinguish behaviors of different instances regarding power state locking.

**Comment by erwango** [comment]
> An example is nRF54H20 in which one instance of MVDMA gets power cut off in 2 sleep states while the remaining instances lose power in 3 sleep states. So the MVDMA driver needs to distinguish behaviors of different instances regarding power state locking.

Then maybe instead of defining all states in all nodes, we can define only the specific states in specific nodes. See https://github.com/zephyrproject-rtos/zephyr/pull/59200/files#diff-5bd998735a6cb86347623d54854157ee8e578ecee7507873c953c18fe4d01d2c

**Comment by ceolin** [comment]
> > An example is nRF54H20 in which one instance of MVDMA gets power cut off in 2 sleep states while the remaining instances lose power in 3 sleep states. So the MVDMA driver needs to distinguish behaviors of different instances regarding power state locking.
> 
> Then maybe instead of defining all states in all nodes, we can define only the specific states in specific nodes. See https://github.com/zephyrproject-rtos/zephyr/pull/59200/files#diff-5bd998735a6cb86347623d54854157ee8e578ecee7507873c953c18fe4d01d2c

I don't think it is the same thing (unless I am missing something). These notifications don't allow you to stop the SoC to sleep, they are sent after the policy has chosen the power state. What this change provides is a way to devices set constraints to forbid the SoC to use specific power states.

**Comment by erwango** [comment]
> I don't think it is the same thing (unless I am missing something). These notifications don't allow you to stop the SoC to sleep, they are sent after the policy has chosen the power state. What this change provides is a way to devices set constraints to forbid the SoC to use specific power states.

I'm not commenting about the notifications, but about what to document in device tree.
In this regard, this is close:
The point is to be able to document in which states a peripheral **supporting PM** will loose context.
For peripheral where PM is not supported, device PM is not coming into play anyway.
Note: we don't need to document states where context is lost by default (standby for instance)

Also one question for @hubertmis: what action do you want to take for the states where MVDMA gets cut off ?
- prevent the system to enter these states ?
- be able to re-initialize MVDMA when you detect that this state is exited ?

**Comment by hubertmis** [comment]
> Also one question for @hubertmis: what action do you want to take for the states where MVDMA gets cut off ?

While MVDMA operation is active, the CPU shall not enter any of the power states cutting power off the active MVDMA instance. Otherwise, interrupted DMA operations could result in data corruption.

So I expect code like this in the MVDMA driver:

```
pm_handling_function() {
  switch (action) {
    case PM_DEVICE_ACTION_SUSPEND:
        /* suspend the device */
        pm_policy_device_power_lock_get(put);
        break;
    case PM_DEVICE_ACTION_RESUME:
        /* resume the device */
        pm_policy_device_power_lock_get(dev);
        break;
}

dma_start() {
  pm_device_runtime_get(dev);
  ...
}

dma_end_handler() {
  pm_device_runtime_put(dev);
  ...
}
```

**Comment by ceolin** [comment]
> > I don't think it is the same thing (unless I am missing something). These notifications don't allow you to stop the SoC to sleep, they are sent after the policy has chosen the power state. What this change provides is a way to devices set constraints to forbid the SoC to use specific power states.
> 
> I'm not commenting about the notifications, but about what to document in device tree. In this regard, this is close: The point is to be able to document in which states a peripheral **supporting PM** will loose context. For peripheral where PM is not supported, device PM is not coming into play anyway.

That is not true, even if the device does not implement "device pm" it still may need to prevent the system to sleep when it is doing something.

> 
> Also one question for @hubertmis: what action do you want to take for the states where MVDMA gets cut off ?
> 
> * prevent the system to enter these states ?
> * be able to re-initialize MVDMA when you detect that this state is exited ?

File: dts/bindings/base/pm.yaml
Comment (decsny) [inline dts/bindings/base/pm.yaml:34]: why put the zephyr, prefix here?

File: dts/bindings/base/pm.yaml
Comment (decsny) [inline dts/bindings/base/pm.yaml:37]: just out of curiosity, why did you choose to list the disabling instead of the enabling states

**Review by decsny — COMMENTED** [review]
why is this not something that can be handled somehow through power domains?

File: dts/bindings/base/pm.yaml
Comment (ceolin) [reply dts/bindings/base/pm.yaml:34]: Because this property is Zephyr's specific. https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#the-zephyr-prefix

File: dts/bindings/base/pm.yaml
Comment (ceolin) [reply dts/bindings/base/pm.yaml:37]: Because that is the information that you need. You want to know which states cause power loss to a device so you can put a constraint in these states to forbid the SoC to use them when it is idle.

**Comment by ceolin** [comment]
> why is this not something that can be handled somehow through power domains?

Because power domains don't control the SoC state, power domains are (in a really abstract and superficial way) a cluster of devices. What we want't here is basically to influence the pm policy to not use specific power states that could cause power loss (or something else that compromises the device behavior) when the device is in use.

File: dts/bindings/base/pm.yaml
Comment (decsny) [reply dts/bindings/base/pm.yaml:34]: do we share the pm binding with an upstream linux binding?

Comment (ceolin) [reply dts/bindings/base/pm.yaml:34]: some properties ... `min-residency-us` and `exit-latency-us`

Comment (decsny) [reply dts/bindings/base/pm.yaml:34]: is the pm binding in zephyr supposed to be analogous to the `domain-idle-states` linux binding?

Comment (ceolin) [reply dts/bindings/base/pm.yaml:34]: there are similarities / inspiration. But we have it tailored to Zephyr needs

File: tests/subsys/pm/power_states_api/src/main.c
Comment (ceolin) [reply tests/subsys/pm/power_states_api/src/main.c:95]: not really needed. The ongong_operation ensures that if suspend to ram is used an assert will happen in pm_state_set. But it is clear having it here, I will add it.

**Comment by ceolin** [comment]
@hubertmis would mind take another look ?

File: dts/bindings/base/pm.yaml
Comment (decsny) [reply dts/bindings/base/pm.yaml:34]: That's why I am wondering why we need the zephyr, prefix here, if the binding itself is zephyr specific

Comment (decsny) [reply dts/bindings/base/pm.yaml:34]: nevermind, this is the base pm.yaml, I thought it was a compatible binding, I read wrong.

File: dts/bindings/base/pm.yaml
Comment (erwango) [inline dts/bindings/base/pm.yaml:37]: Should this encompass states in which devices lose power **by definition** ?

I already had a discussion on this point [here](https://github.com/zephyrproject-rtos/zephyr/pull/60369#discussion_r1305436477). By definition of s2ram, most devices will lose power in this state.

To me `zephyr,disabling-power-states` should only document s2idle states.
Describing in a device node that it is loosing power in standby state doesn't make sense to me

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/base/pm.yaml
Comment (hubertmis) [reply dts/bindings/base/pm.yaml:37]: > By definition of s2ram, most devices will lose power in this state.

"most" is the reason for creating this. Assuming a soc has 3 s2ram substates defined, a device can lose power in 2 of them. Another device can lose power in one, while yet another instance of the same device in all 3. We need a way to clearly define what set of devices is covered by "most" and act accordingly in the device drivers.

One scenario we need to cover is nRF54H20 where almost all MVDMA instances lose power in s2ram substates 0 and 1, but there is one instance that loses power in s2ram substate 1 only. Without this PR we could not use the special MVDMA instance while a CPU is in s2ram. This would unnecessarily limit the available system features.

Comment (erwango) [reply dts/bindings/base/pm.yaml:37]: Then what about combining information from each state and information in specific impacted nodes ?

Note that if this is merged, on STM32, we'll need to add this property in > 1000 device tree nodes, across 10th of device drivers types. With the involved possiblilities of errors.
This is not a minor thing.

Hence I'd like to be really sure that all other ideas have been carefully studied.

Comment (erwango) [reply dts/bindings/base/pm.yaml:37]: Note also, that instead of listing all impacted states, we can mention the "highest" low power state causing this. I haven't seen cases where clocks/power could be lost in one state and preserved in the next lower low power state

**Review by erwango — DISMISSED** [review]

File: dts/bindings/base/pm.yaml
Comment (hubertmis) [reply dts/bindings/base/pm.yaml:37]: > Note that if this is merged, on STM32, we'll need to add this property in > 1000 device tree nodes, across 10th of device drivers types. With the involved possiblilities of errors. This is not a minor thing.

I understand the new property is optional. If you don't intend to use the framework introduced in this PR, you don't need to update the device tree nodes. When you decide to start using this framework, you can start updating the device drivers one by one, each modification associated with updating all compatible device tree nodes.

> Note also, that instead of listing all impacted states, we can mention the "highest" low power state causing this. I haven't seen cases where clocks/power could be lost in one state and preserved in the next lower low power state

I agree. I also haven't seen anything like this. Moreover, if we used only the "highest" power state in the device tree, it would make adding power states easier (fewer device tree nodes to update).

File: dts/bindings/base/pm.yaml
Comment (hubertmis) [reply dts/bindings/base/pm.yaml:34]: I don't understand how the device drivers could use these details. When a device driver is activated, it locks all power states, resulting in clock loss (including those with loss of registers). When a device driver is deactivated, it unlocks all power states (including these with loss of registers). When the device driver gets activated again, it has no information on which power states were entered meantime, so it must assume the worst - the registers' values were lost and must be restored.

Comment (erwango) [reply dts/bindings/base/pm.yaml:34]: >  When the device driver gets activated again, it has no information on which power states were entered meantime, so it must assume the worst - the registers' values were lost and must be restored.

If registers where not lost we should not take time to restore them at resumption. Restoring registers take time it should not be done by default.
This is why we have various power states with min-residency-us and exit-latency-us for each state. Depending on the deepness of the power state, more actions have to be taken at resumption. Based on these information, PM subsystem arbitrate the next sleep state to enter depending on the idle time available.

Comment (hubertmis) [reply dts/bindings/base/pm.yaml:34]: I agree that we should avoid time-consuming restoring if possible. But I don't see from where a device driver could take the information if restoring is needed. Do you have a proposal on how to modify the framework to provide this information to the drivers?

Comment (erwango) [reply dts/bindings/base/pm.yaml:34]: See https://github.com/zephyrproject-rtos/zephyr/pull/69249#issuecomment-1971368367

Comment (hubertmis) [reply dts/bindings/base/pm.yaml:34]: Fair enough. The notification function misses the substate information, but I think it can be fixed.
With this context, I agree more detailed properties in the device tree nodes would be useful. Right now I'm mostly interested in the property describing power states resulting in the loss of clock (not necessarily register values).

Comment (ceolin) [reply dts/bindings/base/pm.yaml:34]: That is something I will fix. The problem that this pr address is not achievable with notifications because the notification is sent after the policy has chosen which state the SoC will use. This pr allows a device set constraints in these states to avoid them to be used.

File: dts/bindings/base/pm.yaml
Comment (ceolin) [reply dts/bindings/base/pm.yaml:37]: > > Note that if this is merged, on STM32, we'll need to add this property in > 1000 device tree nodes, across 10th of device drivers types. With the involved possiblilities of errors. This is not a minor thing.
> 
> I understand the new property is optional. If you don't intend to use the framework introduced in this PR, you don't need to update the device tree nodes. When you decide to start using this framework, you can start updating the device drivers one by one, each modification associated with updating all compatible device tree nodes.
> 
> > Note also, that instead of listing all impacted states, we can mention the "highest" low power state causing this. I haven't seen cases where clocks/power could be lost in one state and preserved in the next lower low power state
> 
> I agree. I also haven't seen anything like this. Moreover, if we used only the "highest" power state in the device tree, it would make adding power states easier (fewer device tree nodes to update).

What "highest" means here ? The deepest state that the SoC can go without loosing power ?

Comment (hubertmis) [reply dts/bindings/base/pm.yaml:37]: > > > Note also, that instead of listing all impacted states, we can mention the "highest" low power state causing this. I haven't seen cases where clocks/power could be lost in one state and preserved in the next lower low power state
> > 
> > 
> > I agree. I also haven't seen anything like this. Moreover, if we used only the "highest" power state in the device tree, it would make adding power states easier (fewer device tree nodes to update).
> 
> What "highest" means here ? The deepest state that the SoC can go without loosing power ?

The most natural way of expressing this seems to be the lightest sleep that disables power (or clock) of a device. And a way for a device driver to lock this and all deeper sleep sub-states and states.

**Comment by mmahadevan108** [comment]
@ceolin, thanks for the PR. Overall the changes look fine. Can you please update the documentation at the below link as well.
https://docs.zephyrproject.org/latest/services/pm/system.html

**Comment by ceolin** [comment]
> @ceolin, thanks for the PR. Overall the changes look fine. Can you please update the documentation at the below link as well. [docs.zephyrproject.org/latest/services/pm/system.html](https://docs.zephyrproject.org/latest/services/pm/system.html)

Absolutely, thanks for reviewing it.

**Comment by erwango** [comment]
@ceolin As already mentioned, what bothers me with this PR is the need to document this property in all devices nodes, while in most cases, the behavior is just "by definition" of the power state.
What about adding a property in power states nodes, which will provide the default behavior applied in general and then this behavior could be amended upon use of the use of ` zephyr,disabling-power-states` device property on devices that have a specific behavior vs the default power state definition ?

**Comment by ceolin** [comment]
> @ceolin As already mentioned, what bothers me with this PR is the need to document this property in all devices nodes, while in most cases, the behavior is just "by definition" of the power state. What about adding a property in power states nodes, which will provide the default behavior applied in general and then this behavior could be amended upon use of the use of ` zephyr,disabling-power-states` device property on devices that have a specific behavior vs the default power state definition ?

@erwango that is optional. Only devices that need to set constraints need to use it. I see what you mean, and that is true for most scenarios. But depending on the topology, a devices may be in different power rails and different power states can turn off different power rails. Keeping it close to the device provides a fine grain that covers all cases (at least it should).

If your concern is due resources needed, we can optimize it, keeping this information out of the device struct and only devices that need it will create it. Similar to what we do with the pm struct. But honestly, this can be addressed later. For now that is optional, as you can see there is no change in any device and consequently no changes in resources consumption.

**Comment by erwango** [comment]
> @erwango that is optional. Only devices that need to set constraints need to use it. I see what you mean, and that is true for most scenarios. But depending on the topology, a devices may be in different power rails and different power states can turn off different power rails. Keeping it close to the device provides a fine grain that covers all cases (at least it should).

Ok for this concern.

Would it then be possible to address following other comments:
 - [Misleading property naming](https://github.com/zephyrproject-rtos/zephyr/pull/70111#discussion_r1604668958): "disabling-power-states"
 - [Only mention the lightest impacted state](https://github.com/zephyrproject-rtos/zephyr/pull/70111#discussion_r1604699400): the deeper sleeping states have the same effect on the device. Listing all states is superfluous.

**Comment by ceolin** [comment]
> > @erwango that is optional. Only devices that need to set constraints need to use it. I see what you mean, and that is true for most scenarios. But depending on the topology, a devices may be in different power rails and different power states can turn off different power rails. Keeping it close to the device provides a fine grain that covers all cases (at least it should).
> 
> Ok for this concern.
> 
> Would it then be possible to address following other comments:
> 
> * [Misleading property naming](https://github.com/zephyrproject-rtos/zephyr/pull/70111#discussion_r1604668958): "disabling-power-states"

I don't exactly how to answer it generically since we don't have this information anywhere today, I mean no power state  tells you it. From device perspective (when the soc goes to listed states) it is the same of turn off action. Each device may interpret it as it needs. That is one of the reason it is optional.

> * [Only mention the lightest impacted state](https://github.com/zephyrproject-rtos/zephyr/pull/70111#discussion_r1604699400): the deeper sleeping states have the same effect on the device. Listing all states is superfluous.

It can be done, but the cost is making it less flexible and transfer some logic to the policy. I rather to have it generic and then based on more usage see if we can generalize it.

**Review by hubertmis — DISMISSED** [review]

**Comment by ceolin** [comment]
@hubertmis thanks for the review. I've to rebase it. Would you mind to refresh your approval ?

**Comment by ceolin** [comment]
@erwango can you take another look please ?

**Comment by erwango** [comment]
Removing my blocker, but not approving as I stand on my last comments. See https://github.com/zephyrproject-rtos/zephyr/pull/70111#issuecomment-2141565699

**Comment by ceolin** [comment]
> Removing my blocker, but not approving as I stand on my last comments. See [#70111 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/70111#issuecomment-2141565699)

Fair enough, we can revisit it later, it is not written in stone. It has valid use cases now and if we find a simpler way to do it after using it we can always change :)

**Review by hubertmis — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]
Thanks, looks good.

**Comment by gmarull** [comment]
I realize now that this PR pollutes struct device for no good reason, introducing a questionable design: introducing PM code in device.h (again). This was merged without any approval from device subsystem maintainers, so I think it should be reverted or provide an urgent fix. We have for a reason a structure for PM resources in struct device, passed as a pointer when devices are initialized. Please provide some feedback, or I'll send a revert PR.

**Comment by gmarull** [comment]
FYI, https://github.com/zephyrproject-rtos/zephyr/pull/74967

**Comment by ceolin** [comment]
> I realize now that this PR pollutes struct device for no good reason, introducing a questionable design: introducing PM code in device.h (again). This was merged without any approval from device subsystem maintainers, so I think it should be reverted or provide an urgent fix. We have for a reason a structure for PM resources in struct device, passed as a pointer when devices are initialized. Please provide some feedback, or I'll send a revert PR.

You were assigned to review it since the beginning and discussions happened for three months, it was plenty of time.
The revert pr was sent at same time you commented here, that is not like giving any time for feedback :|

I have commented there and can comment here, this is not device power management and that is why it was not added in the device pm code. I don't think removing it is fair, honestly. There is a reason for the feature, the implementation can be changed but remove the whole feature this way is unacceptable.

**Pull request merged**
